File: leetcode/0054-spiral-matrix/solution.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        ArrayList<Integer> ans = new ArrayList<>();
        int rs=0,re=matrix.length-1;
        int cs=0,ce=matrix[0].length-1;
        int i=0;
        
        while(rs <= re && cs <= ce){
            if (rs <= re){
                for(int j=cs;j<=ce;j++){
                    ans.add(matrix[rs][j]);
                }
                rs++;
            }

            if (cs <= ce){
                for(int j=rs;j<=re;j++){
                    ans.add(matrix[j][ce]);
                }
                ce--;
            }

            if (rs <= re){
                for(int j=ce;j>=cs;j--){
                    ans.add(matrix[re][j]);
                }
                re--;
            }

            if (cs <= ce){
                for(int j=re;j>=rs;j--){
                    ans.add(matrix[j][cs]);
                }
                cs++;
            }

        }

        return ans;
    }
}

File: leetcode/0054-spiral-matrix/solution.py
class Solution(object):
    def spiralOrder(self, matrix):
        m = len(matrix)
        n = len(matrix[0])

        ans = []

        lc,rc,tr,br = 0,n-1,0,m-1

        while lc <= rc:
            for i in range(lc,rc+1):
                ans.append(matrix[tr][i])
            tr += 1
            print(ans)

            if tr>br:
                break
            for i in range(tr,br+1):
                ans.append(matrix[i][rc])
            rc -= 1

            if lc>rc:
                break
            for i in range(rc, lc-1, -1):
                ans.append(matrix[br][i])
            br -= 1

            if tr>br:
                break
            for i in range(br, tr-1, -1):
                ans.append(matrix[i][lc])
            lc += 1

        return ans
        
